### DISCUSSION TITLE
💡QChatGPT 插件 API 设计需求讨论

### DISCUSSION THREAD

**Discussion by RockChinQ** [langbot-app/LangBot#637] [Ideas]
QChatGPT 预计在下一个版本设计新的统一的 插件API 用于保证版本更新时的插件兼容性。欢迎插件开发者在此处回复。
希望收到以下回复信息：

- 需求的插件 API （如：获取 QChatGPT 版本号）
- 插件系统设计上的任何改进建议

**Comment by RockChinQ** [comment]
插件 API 拟定以 HTTP 方法类型作为标准，此回复中列出的 API 均为目前考虑实现的 API：

## 环境信息

GET QChatGPT 版本
GET 断言 QChatGPT 版本

## 消息处理

POST 发送私聊/群聊消息

## 插件相关

## 会话处理
## 大模型交互
## 数据收集
## 配置项

GET 加载配置文件
POST 修改并保存配置文件

**Comment by zuo-shi-yun** [comment]
- api
    - 获得配置
    - 更改配置。我现在通过pkg.qqbot.cmds.system.cconfig包下的config_operation修改配置
    - 发送信息。现在通过PluginHost发送信息。同时希望把发送群聊信息、私聊信息统一成一个方法，通过launcher_type来判断信息类型
    
- 一些问题
    -  项目的pkg包在插件开发时需要额外引入并且添加到gitingore文件中，十分累赘，同时pkg包更新无法及时获得，更新起来也挺费劲，可否封装成pypl包？
    - 插件开发时导入本插件其他包时引用的路径为plugins开头，如plugins.hello，但是本地开发时是没有plugins这个文件夹的，这就导致每次开发若想使用自动导入等静态检查的话需要在本地新建plugins文件夹，同时把项目文件复制一份进去，同样需要添加gitingore，每次更新同样很费劲。这个问题不知如何解决
    - 添加一个统一的回复信息事件。现在只有一个NormalMessageResponded事件，当api超时(网络风暴)或者被命令触发的回复均无法触发NormalMessageResponded事件
    - 添加一个统一的插件初始化函数。现在通过在main函数的init中初始化，结构看起来怪怪的。

**Reply by RockChinQ** [reply]
## API

- 配置项相关 API 确认加入
- 消息发送 API 还会详细设计一下

## 插件开发方式

开发方式目前比较推荐的步骤是：

1. Clone 主程序
2. 在 `plugins` 目录下新建 Git 目录
3. 编写插件
4. 启动主程序调试
5. 发布插件

可以考虑一下这种开发方式。  
把 QChatGPT 发行到 PyPI、发布插件开发 SDK 之类的工作比较繁琐，近期可能不会考虑。

## 其他

- 这个事件会考虑增加，如果需要更多的事件钩子，可以继续回复需求
- 之后会把插件与主程序之间的所有交互（包括注册、加载）全部封装到插件 API

**Comment by firedrakeovo** [comment]
* 需求插件：  bot的好友和群聊申请相关功能
* 设计建议： 1.可以添加管理员cmd获取当前bot账号待同意的好友和群聊申请列表，选择对应项进行同意或拒绝
&emsp;&emsp;&emsp;&emsp;&ensp;&ensp; 2.通过管理员cmd让bot账号主动对群聊或好友进行添加申请。

**Comment by RockChinQ** [comment]
### 插件系统相关

- 获取插件列表
- 获取插件启用状态
- 获取插件注册的事件列表
